### ISSUE TITLE
Handsontable colHeaders gets NaN on loadData

### ISSUE THREAD

**Issue by tkane2000** [handsontable/handsontable#5369]
### Description
With the following options set:
- data: array of one object that doesn't contain a property for each column.  for instance, the table may have 10 columns, but the object only has 6 properties.
- colHeaders: function
- columns: array of objects
- manualColumnMove: array of numbers

When you call hot.loadData(null), the colHeaders callback is called with NaN for some of the columns.

Note: calling hot.loadData with an empty array has similar, but different issues.

### Steps to reproduce
(you're best bet is to take a look at the fiddle I put together)
1. Set the following options:
- data: array of one object that has 6 properties.
- colHeaders: function(col) { return '<span>' + col + '</span>' }
- columns: array of 10 objects
- manualColumnMove: array of numbers, length 10
2. Instantiate your table
3. call loadData with null

Note: column headers are incorrect.

### Demo
https://jsfiddle.net/tkane2000/tyh8qxve/

### Research
It seems the problem arises from the following (pseudo) call stack:
ManualColumnMove.updateColumnsMapper
 -> hot.countSourceCols()

hot.countSourceCols derives the count from instance.getSourceData, which counts the properties in the first object (row) of the hot data.  If this object doesn't have as many properties as there are columns, then you get the issue.

As a solution, I created a patch, which I'll post below, that overrides getSourceData to return the length of the columns.  That seems to fix the issue, but I'm not sure what the initial intention of countSourceCols is and if it could break anything.  (Please advise)

If manualColumnMove is set to true instead of an array the problem doesn't arise

### Patch
```
function patchHOTCountSourceCols(tableInstance) {
  if (tableInstance._originalCountSourceCols) {
    return;
  }

  tableInstance._originalCountSourceCols = tableInstance.countSourceCols;

  tableInstance.countSourceCols = function countSourceCols() {
    var len = 0;
    var settings = tableInstance.getSettings();

    if (settings && settings.columns) {
      len = settings.columns.length;
    } else {
      len = tableInstance._originalCountSourceCols();
    }

    return len || 0;
  };
}
```

### Your environment
* Handsontable version: 3.0.0
* Browser Name and version: Chrome
* Operating System: Mac OS X

**Comment by AMBudnik** [comment]
Hi @tkane2000 

just to confirm, to get the issue I click the <kbd>loadData: []</kbd> and then I call <kbd>loadData: data</kbd>? As the <kbd>loadData: data</kbd> works well if it's the first button be to pushed.

Ps. I've tried the v5.0.1 (latest) as well but if the scenario doesn't change the result stays the same.

**Comment by tkane2000** [comment]
Hi @AMBudnik, thanks for getting back to me!

The scenario you mentioned will cause issues.  The other scenario is to just press "loadData:null" and this is what we ran into in our app.  (be sure to re-run fiddle between tests)

You'll see NaN in the column headers.  Take a look at the colHeaders callback and you'll see where those values are coming from.

Lastly, if you uncomment this line, it seems to fix everything:
// patchHOTCountSourceCols(hot1);
That's the solution I plan to use locally, but was hoping you could weigh in and let me know if it could cause any problems.

**Comment by AMBudnik** [comment]
Hi @tkane2000 

It looks like some settings are interfering with the colHeaders attachment.

I've simplified the demo https://jsfiddle.net/mzyLhn38/

and then I no longer get the issue in line` 6-7`
```
1. click "loadData: null"
result: last 4 col headers are "NaN"
```

![1gif](https://user-images.githubusercontent.com/10757813/44511264-3c8e8800-a6b7-11e8-913e-6f4db5e7211d.gif)

But... when I do the scenario at` 9-11` There's still

> result, cols 1-4 are "NaN"

**Comment by AMBudnik** [comment]
I haven't found any solution for this issue. It looks like we need to make some changes in the source code.

**Comment by tkane2000** [comment]
Thanks @AMBudnik!  Unfortunately, we require showing columns even if they don't contain data, so removing the "columns" setting isn't an option. :\

What do you think about the solution in the patch, which changes countSourceCols to return settings.columns.length?

**Comment by AMBudnik** [comment]
It is really hard to say. It is a really old part of the code and without an investigation, it would hard to say anything. Surely there's some kind of mismatch between expected data and the one that we get so it may be a wrong hook order as well.

**Comment by tkane2000** [comment]
We’re going to use the patch for now. If you’d like me to open a PR let me know. (Someone would still need to do the investigation)

**Comment by AMBudnik** [comment]
We're pretty much bulked with work right now but a PR with tests is always welcomed :)

**Comment by AMBudnik** [comment]
Hey @tkane2000 

can you help me to test this bug with our new resources? We're working on a new index mapper and are doing some tests.

I've used the resources in the demo from this example and it seems to work well.

https://cdn.jsdelivr.net/gh/handsontable/handsontable@293f30429d37fea2add8a9c3df5917b2684ce02a/dist/handsontable.full.css
https://cdn.jsdelivr.net/gh/handsontable/handsontable@293f30429d37fea2add8a9c3df5917b2684ce02a/dist/handsontable.full.js

Previously, we got `NaN` headers
![old](https://user-images.githubusercontent.com/10757813/65764106-24443480-e125-11e9-9f02-a0fde7d7793b.gif)

And here's after loading new resources https://jsfiddle.net/cfrazx8s/
![new](https://user-images.githubusercontent.com/10757813/65764112-28705200-e125-11e9-9697-320b7dfa4291.gif)

**Comment by tkane2000** [comment]
I’ll take a look!

**Comment by AMBudnik** [comment]
Hi @tkane2000 

we released 8.0.0-beta11 yesterday. Now it is easier to test. You can download it from npm using

`npm i handsontable@beta`

**Comment by tkane2000** [comment]
Thanks Aleksandra!  We're pretty busy right now, but I'll try to get to it!

Best - Trevor

**Comment by tkane2000** [comment]
I took a quick look at your jsfiddle: https://jsfiddle.net/cfrazx8s/
In your gifs it looks like you were able to reproduce the error, but I did it a bit differently by adding in columns, specifically, more column settings than there are properties in the item (where the item is represented by a row in the table and each property of the item is displayed in a corresponding column).
However, I added the columns settings to your fiddle and everything seems to work!
How can I see the diffs for this fix in github?
Also, do you plan to apply this fix to earlier versions?  We're on 3.0.0...upgrading will be painful. :(

**Comment by wojciechczerniak** [comment]
Great to know that it works for you @tkane2000 

>  How can I see the diffs for this fix in github?

All changes are in PR #5945

> Also, do you plan to apply this fix to earlier versions? 

No. Sorry, there were too many changes since 3.0.0

**Comment by AMBudnik** [comment]
Hi @tkane2000 

We've just released the official, stable Handsontable v 8.0.0. Can you confirm that it works for you after updating?

**Comment by AMBudnik** [comment]
The issue seems to be fixed for v 8.0.0.
Please correct me if I'm wrong @tkane2000 and I'll reopen the issue.
